### PR TITLE
[Impeller] Allocate images into host buffer, blit to device private texture.

### PR DESCRIPTION
**Pull request by jonahwilliams**
~~I'm not sure if we can land this until we figure out image upload performance issues~~ 

Patch landed to force serialization of texture upload in https://github.com/flutter/engine/pull/40464

Updated patch so that multiframe codecs always go to host visible textures and don't generate mipmaps. This should help throughput performance since UMA don't have much cost for creating these textures

---------------------
This would allow us to apply lossy compression to image textures, which dramatically reduces (50%?) memory usage on certain formats. From experimenting with the metal memory profiler, this does appear to use less memory that using host visible textures - possibly due to lossless compression. I've also confirmed that the primary memory category we measure does not include private metal resources, so there is an accounting aspect to it. That said, all categories appear lower after this change.

I experimented with this by swiping through wonderous for a few seconds and then letting the app rest a bit:

Using the VM tracker which includes all alloctions, without this patch we stabilize at ~320 MB of dirty memory and 2 MB of swap, while with this patch we stabilize at 237 MB of dirty memory and 42 MB of swap.

The metal resource events category also is reduced with this patch, previously peaking at 325 MB and dropping down to 287. Finally, mostly due to accounting the all heap and anon VM category drops from 280 to 90 MB.

Leaves the existing path more or less for gles/vulkan as we need to determine if private textures make sense there and implement support for buffer to texture blit.

https://github.com/flutter/flutter/issues/122914
https://github.com/flutter/flutter/issues/122918
https://github.com/flutter/flutter/issues/122406

### CODE REVIEW

File: impeller/renderer/backend/gles/blit_pass_gles.h [hunk 1]
```diff
@@ -50,6 +50,15 @@ class BlitPassGLES final : public BlitPass {
                                     size_t destination_offset,
                                     std::string label) override;
 
+  // |BlitPass|
+  bool OnCopyBufferToTextureCommand(std::shared_ptr<DeviceBuffer> source,
+                                    std::shared_ptr<Texture> destination,
+                                    size_t source_offset,
+                                    IPoint destination_origin,
+                                    std::string label) override {
+    return false;
```
Comment (jonahwilliams) [inline impeller/renderer/backend/gles/blit_pass_gles.h:59]: TODO

File: impeller/renderer/backend/vulkan/blit_pass_vk.h [hunk 2]
```diff
@@ -50,6 +50,15 @@ class BlitPassVK final : public BlitPass {
                                     size_t destination_offset,
                                     std::string label) override;
 
+  // |BlitPass|
+  bool OnCopyBufferToTextureCommand(std::shared_ptr<DeviceBuffer> source,
+                                    std::shared_ptr<Texture> destination,
+                                    size_t source_offset,
+                                    IPoint destination_origin,
+                                    std::string label) override {
+    return false;
```
Comment (jonahwilliams) [inline impeller/renderer/backend/vulkan/blit_pass_vk.h:59]: TODO

File: impeller/renderer/backend/metal/allocator_mtl.mm [hunk 3]
```diff
@@ -194,6 +195,11 @@ static MTLStorageMode ToMTLStorageMode(StorageMode mode,
 
   mtl_texture_desc.storageMode = ToMTLStorageMode(
       desc.storage_mode, supports_memoryless_targets_, supports_uma_);
+  if (@available(macOS 12.5, ios 15.0, *)) {
```
Comment (jonahwilliams) [inline impeller/renderer/backend/metal/allocator_mtl.mm:198]: TODO: make configurable, figure out if it has reasonable behavior for image textures

File: lib/ui/painting/image_decoder_unittests.cc [hunk 4]
```diff
@@ -272,23 +272,23 @@ TEST_F(ImageDecoderFixtureTest, ValidImageResultsInSuccess) {
...
-  bool negative = half >> 15;
-  uint16_t exponent = (half >> 10) & 0x1f;
-  uint16_t fraction = half & 0x3ff;
-  float fExponent = exponent - 15.0f;
-  float fFraction = static_cast<float>(fraction) / 1024.f;
-  float pow_value = powf(2.0f, fExponent);
-  return (negative ? -1.f : 1.f) * pow_value * (1.0f + fFraction);
-}
-}  // namespace
+// namespace {
```
Comment (jonahwilliams) [inline lib/ui/painting/image_decoder_unittests.cc:275]: TODO: all of these tests need to be moved to a different TU that has an impeller context

File: impeller/renderer/backend/gles/blit_pass_gles.h (hunk 1 above)
Comment (chinmaygarde) [reply impeller/renderer/backend/gles/blit_pass_gles.h:59]: Totally fine. But `IMPELLER_UNIMPLEMENTED` in `//impeller/base/config.h`

File: impeller/renderer/backend/metal/allocator_mtl.mm [hunk 5]
```diff
@@ -194,6 +195,11 @@ static MTLStorageMode ToMTLStorageMode(StorageMode mode,
...
+    if (mtl_texture_desc.storageMode == MTLStorageModePrivate) {
+      mtl_texture_desc.compressionType = MTLTextureCompressionTypeLossy;
```
Comment (chinmaygarde) [inline impeller/renderer/backend/metal/allocator_mtl.mm:200]: I'd like to attempt this in another patch and have this flag be present on `impeller::TextureDescriptor`.

File: impeller/renderer/backend/gles/blit_pass_gles.h [hunk 6]
```diff
@@ -50,6 +50,15 @@ class BlitPassGLES final : public BlitPass {
                                     size_t destination_offset,
                                     std::string label) override;
 
+  // |BlitPass|
+  bool OnCopyBufferToTextureCommand(std::shared_ptr<DeviceBuffer> source,
```
Comment (chinmaygarde) [inline impeller/renderer/backend/gles/blit_pass_gles.h:54]: I think we should make the source a buffer view. That way we can specify ranges in the source. Also, you'll be able to get rid of source_offset which pretty much does the same thing.

File: impeller/renderer/backend/vulkan/blit_pass_vk.h (hunk 2 above)
Comment (chinmaygarde) [reply impeller/renderer/backend/vulkan/blit_pass_vk.h:59]: No problem. I'll handle it. But again, FML_UNIMPLEMENTED so its easier to grep for more than anything else.

File: impeller/renderer/blit_command.h [hunk 7]
```diff
@@ -28,6 +28,13 @@ struct BlitCopyTextureToBufferCommand : public BlitCommand {
   size_t destination_offset;
 };
 
+struct BlitCopyBufferToTextureCommand : public BlitCommand {
+  std::shared_ptr<DeviceBuffer> source;
```
Comment (chinmaygarde) [inline impeller/renderer/blit_command.h:32]: Consider a buffer view here.

**Review by chinmaygarde — COMMENTED** [review]
Initial comments but all nits. We can rework this later as needed. Let's not worry about the other backends either.

File: impeller/renderer/backend/gles/blit_pass_gles.h (hunk 1 above)
Comment (jonahwilliams) [reply impeller/renderer/backend/gles/blit_pass_gles.h:60]: Done

File: impeller/renderer/backend/vulkan/blit_pass_vk.h (hunk 2 above)
Comment (jonahwilliams) [reply impeller/renderer/backend/vulkan/blit_pass_vk.h:60]: Done

File: lib/ui/painting/image_decoder_unittests.cc (hunk 4 above)
Comment (jonahwilliams) [reply lib/ui/painting/image_decoder_unittests.cc]: Mocked out the allocator. Done

File: impeller/renderer/backend/metal/allocator_mtl.mm (hunk 5 above)
Comment (jonahwilliams) [reply impeller/renderer/backend/metal/allocator_mtl.mm]: Makes sense, removed for now.

File: impeller/renderer/backend/gles/blit_pass_gles.h (hunk 6 above)
Comment (jonahwilliams) [reply impeller/renderer/backend/gles/blit_pass_gles.h]: Done

File: impeller/renderer/blit_command.h (hunk 7 above)
Comment (jonahwilliams) [reply impeller/renderer/blit_command.h]: Done

**Comment by jonahwilliams** [comment]
To make the BufferView change work I had to update the mtl texture to allow casting from a DeviceBuffer to casting from a Buffer.

**Comment by jonahwilliams** [comment]
Landing this as is will likely regress worst frame times on gallery, as texture blitting competes with the frame workload

**Review by chinmaygarde — APPROVED** [review]

**Pull request merged**